### PR TITLE
Use symlinks and hashes to version control images

### PR DESCRIPTION
**Pull request by stefwalter**
Images are simple symlinks in the test/images directories, and
are committed to git and versioned. These symlinks are relative
and point to qcow2 files in the same directory.

The actual image qcow2 data files are named after a hash of
their contents. These are the files that are not committed to
git (too big) but are uploaded to our image store.

After someone changes the setup script, they must:

 * Recreate the images with ./vm-create
 * Commit the generated symlinks
 * Use vm-upload to push the image data to our image store
 * Push the resulting commits

### CODE REVIEW

**Comment by stefwalter** [comment]
Remaining to be done:

[ ] Commit initial symlinks
[ ] vm-download --prune option
[ ] compress uploads and downloads

**Comment by stefwalter** [comment]
Initial symlinks are not hashes, and point to the old numbers for compatibility

**Comment by dperpeet** [comment]
This is a pretty significant change from our previous way of doing things: now all images have to be in the cockpit directory. Before, we used the environment variable `TEST_DATA`.

Why not store the image filename in a text file instead of using a symlink? Then we can look for that in the intended directory. This also takes care of the case of multiple cockpit checkouts.

File: test/vm-download [hunk 1]
```diff
@@ -18,78 +18,79 @@
...
+parser.add_argument("--prune", action="store_true", help="Remove unused images")
+parser.add_argument('image', nargs='*')
+args = parser.parse_args()
+
+def download(image):
+    link = os.path.join(IMAGES, image)
+    if not os.path.islink(link):
+        parser.error("image link does not exist: " + image)
+
+    if not args.force and os.path.exists(dest):
```
Comment (dperpeet) [inline test/vm-download]: `dest` is declared after this use

File: test/vm-download [hunk 2]
```diff
@@ -18,78 +18,79 @@
...
+
+import argparse
+import os
+import subprocess
+import shutil
+import sys
+import tempfile
+
+BASE = os.path.dirname(__file__)
+IMAGES = os.path.join(BASE, "images")
```
Comment (dperpeet) [inline test/vm-download]: why change this from using TEST_DATA? https://github.com/cockpit-project/cockpit/pull/3037/files#diff-82bca57c3ec28019b40b2e8fcd72cb92L26

File: test/testvm.py [hunk 3]
```diff
@@ -83,22 +83,10 @@ def __init__(self, address=None, flavor=None, system=None, arch=None, verbose=Fa
...
 
-        self.tag = "0"
-        tags = self.getconf('tags')
-        if tags and self.os in tags:
-            self.tag = tags[self.os]
-
-        self.image = "%s-%s-%s-%s" % (self.flavor, self.os, self.arch, self.tag)
+        self.image = "%s-%s-%s" % (self.flavor, self.os, self.arch)
         self.test_dir = os.path.abspath(os.path.dirname(__file__))
         self.test_data = os.environ.get("TEST_DATA") or self.test_dir
```
Comment (dperpeet) [inline test/testvm.py]: this conflicts with the change to using the test directory

**Comment by stefwalter** [comment]
Good point about ```TEST_DATA``` will fix to use that.

symlinks are text files when stored in git. They're just very useful text files that the operating system can dereference for us.

**Comment by stefwalter** [comment]
Oh by the way, in case it wasn't obvious ... having CI generate test images was a nice idea, but ended up being a fail in the end ... especially numbered ones. The images need to be generated first, and then their identifier (in this case a sha hash) should be included in the commit. This allows us to iterate on images properly.

**Comment by stefwalter** [comment]
Testing this further, have fixed all reviewed issues locally. Will push when my tree is free.

**Comment by stefwalter** [comment]
#3033 is the test bed for running this through the whole ./vm-create and ./vm-upload process.

**Comment by stefwalter** [comment]
The current copy of image 'cockpit-fedora-23-x86_64-7' online is corrupted. I'll be uploading my local one (from before the outage). But that shouldn't affect testing or review of this on other operating systems.

**Comment by stefwalter** [comment]
Removed last few stragglers of --revision.

File: test/testvm.py [hunk 4]
```diff
@@ -674,22 +658,28 @@ def save(self):
...
             files = [ ]
             # Copy image via convert, to make it sparse again
             files.append(self._image_image)
-            subprocess.check_call([ "qemu-img", "convert", "-O", "qcow2", self._image_image, self._image_original ])
-            # Copy additional ISO as well when it exists
-            with open(self._checksum_original, "w") as f:
-                subprocess.check_call([ "sha256sum" ] + map(os.path.basename, files),
-                                      cwd=self._images_dir,
-                                      stdout=f)
+            partial = self._image_original + ".partial"
```
Comment (dperpeet) [inline test/testvm.py]: very good change using ".partial" here - manual deletion on failure was very annoying

**Comment by stefwalter** [comment]
Bug here:

```
vm-download: error: image link does not exist: cockpit-fedora-22-x86_64
```

File: test/testvm.py (hunk 3 above)
Comment (dperpeet) [inline test/testvm.py]: with the new files in `test/images`, we need to get the filenames from that directory - if `TEST_DATA` is set to something else in the environment, the current decision logic won't be able to find the original image

We get the link / image to use from the git repo, which shouldn't be written into the `TEST_DATA` directory.

Comment (stefwalter) [inline test/testvm.py]: Aha, yes that's a problem. Will give it some thought.

**Comment by stefwalter** [comment]
Added full TEST_DATA support. Will remove needswork once I've tested it.

**Comment by dperpeet** [comment]
looks good!

**Comment by stefwalter** [comment]
Squashed.

**Comment by stefwalter** [comment]
> The current copy of image 'cockpit-fedora-23-x86_64-7' online is corrupted. I'll be uploading my local one (from before the outage). But that shouldn't affect testing or review of this on other operating systems.

Done.

**Pull request closed without merging**